### PR TITLE
Issue 26581/per partition cursor

### PR DESCRIPTION
**Pull request by maxi297**
## What
Addresses https://github.com/airbytehq/airbyte/issues/26581. Note that this is in draft since we need to complete https://github.com/airbytehq/airbyte/issues/26607 before merging

## How
Adding a new cursor class to support per partition states.

TODO:
* Validate if requests performed are identical from master for zenloop (records and slice logs have been validated). If possible/worth, move this to an integration test
* Validate how this interacts with custom components: they have their own state and we should probably instantiate PerPartitionCursor only if the incremental is not a custom component to preserve backward compatibility. This would be only for backward compatibility reasons though and therefore, maybe we need to rely on the format of the stream state to validate this...
* [tentative] Clean up cursor/stream slicer interface if possible: it might be tricky since `SimpleRetriever` thinks stream slicers have `update_cursor` has a method but it does not make sense for ListPartitionRouter and SubstreamPartitionRouter

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py`
2. `airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py`
3. `airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py`

## 🚨 User Impact 🚨
Until https://github.com/airbytehq/airbyte/issues/26607 is completed, this change is not backward compatible and

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 1]
```diff
@@ -389,10 +389,23 @@ def read_records(
...
+        # * Why is this abstraction not on the DeclarativeStream level? DeclarativeStream does not have a notion of stream slicers. However,
+        #    it might make more sense to have this on the stream level since this logic would probably apply to the other retrievers. As of
+        #    2023-06-09, this is a breaking change because it would mean adding `stream_slicer` to the Retriever interface which we would
+        #    to avoid for now.
+        slice_state = self.stream_slicer.select(stream_slice, stream_state) if hasattr(self.stream_slicer, "select") else stream_state
         records_generator = self._read_pages(
             self.parse_records,
             stream_slice,
-            stream_state,
+            slice_state,
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:408]: Does that means that I need to rename all downstream classes from `stream_state` to `slice_state`? I feel like I should but it seems non trivial so I would like to validate before doing so

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py [hunk 2]
```diff
@@ -508,10 +512,15 @@ def _merge_stream_slicers(self, model: DeclarativeStreamModel, config: Config) -
                 else self._create_component_from_model(model=stream_slicer_model, config=config)
             )
 
-        if incremental_sync and stream_slicer:
-            return CartesianProductStreamSlicer(stream_slicers=[incremental_sync, stream_slicer], parameters=model.parameters)
-        elif incremental_sync:
-            return incremental_sync
+        if model.incremental_sync and stream_slicer:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:515]: I need to validate this behavior with custom component. As is, I'm almost certain this would break their state since it would migrate it to the new state per partition. This might mean that:
* https://github.com/airbytehq/airbyte/issues/27024 is needed to merge this
I would also like to review the sources in our repository to make sure there are no other sources than zenloop that has incremental sync and partition routing

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py (hunk 1 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:408]: Actually, I think this is the same concern as https://github.com/airbytehq/airbyte/pull/27223/files#diff-574e7f982441af95b199fa0d628677b9a654543eaea1534b7516c8679375a235R392 and maybe we should wait...

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        self._partition_router = partition_router
+        self._cursor_per_partition = {}
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[PerPartitionStreamSlice]:
+        slices = self._partition_router.stream_slices(sync_mode, self._NO_STATE)
+        for partition in slices:
+            cursor = self._cursor_per_partition.get(self._to_tuple(partition))
+            if not cursor:
+                cursor = self._create_cursor(self._NO_CURSOR_STATE)
+                self._cursor_per_partition[self._to_tuple(partition)] = cursor
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:117]: There is an interesting side-effect of having this method be a generator: if get_stream_state gets called before generator from get_stream_state is exhausted, the response state will be partial. I'm not sure if this is a concern though...

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+            if not cursor:
+                cursor = self._create_cursor(self._NO_CURSOR_STATE)
+                self._cursor_per_partition[self._to_partition_key(partition)] = cursor
+
+            for cursor_slice in cursor.stream_slices(sync_mode, self._get_state_for_partition(partition)):
+                yield PerPartitionStreamSlice(partition, cursor_slice)
+
+    def update_cursor(self, stream_slice: PerPartitionStreamSlice, last_record: Optional[Record] = None):
+        if not last_record:
+            # The `update_cursor` method is called without `last_record` in order to set the initial state. In that case, stream_slice is
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:135]: should this union type be reflected in the method's signature?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+            if cursor_state:
+                states.append(
+                    {
+                        "partition": self._to_dict(partition_tuple),
+                        "cursor": cursor.get_stream_state(),
+                    }
+                )
+        return {"states": states}
+
+    def _get_state_for_partition(self, partition: Mapping[str, Any]) -> Any:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:169]: nit: doesn't this always return a `StreamState`?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py [hunk 6]
```diff
@@ -389,10 +389,23 @@ def read_records(
...
+        # partition. We have generified the problem to the stream slice instead of the partition because it is the level of abstraction
+        # streams know (they don't know about partitions). However, we're still unsure as how it will evolve since we can't see any other
+        # cursor doing selection per slice. We don't want to pollute the interface. Therefore, we will keep the `hasattr` hack for now.
+        # * What is the information we need to clean the hasattr? Once we will have another case where we need to select a state, we will
+        #    know if the abstraction using `stream_slice` so select to state is the right one and validate if the interface makes sense.
+        # * Why is this abstraction not on the DeclarativeStream level? DeclarativeStream does not have a notion of stream slicers. However,
+        #    it might make more sense to have this on the stream level since this logic would probably apply to the other retrievers. As of
+        #    2023-06-09, this is a breaking change because it would mean adding `stream_slicer` to the Retriever interface which we would
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py:402]: I don't see any custom implementation of Retriever. I was thinking about changing this in a later PR. Based on our discussions, we should not rely on `stream_state` as an argument of the method. This mean that we can probably remove `stream_state` as a parameter of this method and use the stream_slicer to get the information. I'll update the comment and have a follow-up PR

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 4 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:135]: Isn't the Optional reflecting that? In any case, I would like to clean this in a follow-up PR and remove the fact that `last_record` is optional

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                raise KeyError(
+                    f"Partition {str(exception)} could not be found in current state based on the record. This is unexpected because "
+                    f"we should only update cursor for partition that where emitted during `stream_slices`"
+                )
+
+    def _init_state(self, stream_state: dict) -> None:
+        if not stream_state:
+            return
+
+        for state in stream_state["states"]:
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:153]: We can but agreed with Sherif that it wouldn't be the best use of our time as we can do a reset on the source and re-sync. For more information about zenloop, check https://github.com/airbytehq/airbyte/pull/27243

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 5 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:169]: `Optional[StreamState]`, yes. Updated!

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 4 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:143]: wrong line. I was referring to
>  In that case, stream_slice is not a PerPartitionStreamSlice but is a dict representing the state

**Review by girarda — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        if type(other) == PerPartitionStreamSlice:
+            # noinspection PyProtectedMember
+            return self._partition == other._partition and self._cursor_slice == other._cursor_slice
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class CursorFactory:
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:88]: Why is this needed?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    _NO_CURSOR_STATE = {}
+    _KEY = 0
+    _VALUE = 1
+
+    def __init__(self, cursor_factory: CursorFactory, partition_router: StreamSlicer):
+        self._cursor_factory = cursor_factory
+        self._partition_router = partition_router
+        self._cursor_per_partition = {}
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[PerPartitionStreamSlice]:
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:127]:
```suggestion
    def stream_slices(self, sync_mode: SyncMode, _: StreamState) -> Iterable[PerPartitionStreamSlice]:
```
WDYT about obscuring the argument name to discourage using it?

**Review by clnoll — CHANGES_REQUESTED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+#
+
+from typing import Any, Iterable, Mapping, Optional
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.declarative.stream_slicers.stream_slicer import StreamSlicer
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice, StreamState
+
+
+class Hashabledict(dict):
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: Those are good points! I gave it a shot with the flattened tuples and it addresses the flaws you were mentioning. Let me know what you think

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    """
+
+    def __hash__(self):
+        return hash(self._to_hashable(self))
+
+    def _to_hashable(self, value):
+        if type(value) == list:
+            return (self._to_hashable(item) for item in value)
+        elif type(value) in {dict, Hashabledict}:
+            return frozenset(value), frozenset((self._to_hashable(item) for item in value.values()))
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: Given the new code:

```
>>> PerPartitionKeySerializer().to_partition_key({1: 1, 2: 2}) == PerPartitionKeySerializer().to_partition_key({1: 2, 2: 1})
False
```

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        return value
+
+
+class PerPartitionStreamSlice(StreamSlice):
+    def __init__(self, partition: Mapping[str, Any], cursor_slice: Mapping[str, Any]):
+        self._partition = partition
+        self._cursor_slice = cursor_slice
+        self._stream_slice = dict(partition) | dict(cursor_slice)
```
[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:64]: This was a concern of mine as well but my mind drifted to other problems. I'll raise an error in this case to avoid confusion

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 8 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:118]: On my first iteration, I had a case where I wanted the `CursorFactory` to provide the cursor field. I later found a better solution and scraped that. However, it felt like just passing a `Callable[[], StreamSlicer]` might incur some refactoring if we see the need to add some behavior on `CursorFactory`. Doest that make sense?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+                yield *previous_keys, tuple([PerPartitionKeySerializer.to_partition_key(dictionary) for dictionary in to_serialize])
+            else:
+                yield *previous_keys, tuple(to_serialize)
+        else:
+            yield *previous_keys, to_serialize
+
+    @staticmethod
+    def to_partition(to_deserialize: Any):
+        root = {}
+        for flattened_entry in to_deserialize:
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:41]: I suppose I wouldn't call this `flattened` since the tuples are nested. (I was thinking of a flattened dictionary as `{'a': {'a1': 1, 'a2': 2}, 'b': 2}` => `((("a", "a1"), 1), (("a", "a2"), 2), (("b",), 2))`, but don't see a need to do this as your nested tuple solution seems fine and more closely resembles the dictionary.) How about just `entry`?

**Review by clnoll — APPROVED** [review]
LGTM @maxi297 just a couple of minor comments.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 13 above)
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py:45]: Okay cool, in that case it makes sense to me to keep `flattened_entry` at the root.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: Hmm, agree this needs some thought. One question that I'm getting hung up on is whether this is any different than what we already deal with when we serialize/deserialize the state message as json, since json doesn't have tuples. Do you know whether we're already dealing with this kind of situation?

[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: We could probably serialize this as JSON given that we iron out some details. One thing I see if the order of fields. In JSON, the order of fields is not relevant regarding equality. However, once parse as a string, this is relevant. The solution with tuples solved this by ordering keys. I'm wondering if there would be other things to consider if we were to use JSON

[the remark this answers is not in the document]
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/per_partition_cursor.py]: Does ensuring that tuples aren't part of state feel tractable to you? Is it possible that we already do that, since the state has to be json serializable and tuples are not json serializable?

One thing I'm not totally sure of is whether we're talking about rare and hypothetical scenarios. Do you have an opinion on that given your experience in this area? I feel like if the list/tuple situation is sufficiently rare, restricting it could be okay (especially if there are workarounds).

**Comment by octavia-squidington-iii** [comment]
##  source-zenloop test report (commit [`abb9e1bad0`](https://github.com/airbytehq/airbyte/pull/27223/commits/abb9e1bad082cd821b63115cbc1a49f8e60646e0)) - ❌

⏲️  Total pipeline duration: 188 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-zenloop/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Code format checks                                                    | ✅        |
| Connector package install                                             | ✅        |
| Build source-zenloop docker image for platform linux/x86_64           | ✅        |
| Acceptance tests                                                      | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5334186103)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-zenloop test
```

**Comment by octavia-squidington-iii** [comment]
##  source-zenloop test report (commit [`6e32c10cff`](https://github.com/airbytehq/airbyte/pull/27223/commits/6e32c10cff686e4a8b52aaabe208c66f75843a0b)) - ❌

⏲️  Total pipeline duration: 157 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-zenloop/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Code format checks                                                    | ✅        |
| Connector package install                                             | ✅        |
| Build source-zenloop docker image for platform linux/x86_64           | ✅        |
| Acceptance tests                                                      | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5334188405)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-zenloop test
```

**Comment by octavia-squidington-iii** [comment]
##  source-zenloop test report (commit [`9018b94ba9`](https://github.com/airbytehq/airbyte/pull/27223/commits/9018b94ba9798358e8b35138429af2f117bcadc8)) - ❌

⏲️  Total pipeline duration: 153 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-zenloop/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Code format checks                                                    | ✅        |
| Connector package install                                             | ✅        |
| Build source-zenloop docker image for platform linux/x86_64           | ✅        |
| Acceptance tests                                                      | ❌        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5336485927)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-zenloop test
```

**Review by clnoll — APPROVED** [review]

**Comment by octavia-squidington-iii** [comment]
##  source-zenloop test report (commit [`e363fd6cb8`](https://github.com/airbytehq/airbyte/pull/27223/commits/e363fd6cb8a705aa60ab2ad3b2b5f51df6acfb6b)) - ❌

⏲️  Total pipeline duration: 153 seconds

| Step                                                                  | Result   |
|:----------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-zenloop/metadata.yaml | ✅        |
| Connector version semver check.                                       | ✅        |
| Connector version increment check.                                    | ❌        |
| QA checks                                                             | ✅        |
| Code format checks                                                    | ✅        |
| Connector package install                                             | ✅        |
| Build source-zenloop docker image for platform linux/x86_64           | ✅        |
| Acceptance tests                                                      | ✅        |

🔗 [View the logs here](https://github.com/airbytehq/airbyte/actions/runs/5336654627)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-zenloop test
```
